token-lending/program/src/processor.rs

```diff
@@ -575,10 +675,36 @@ fn process_borrow(
         }
     }
 
+    let mut obligation = Obligation::unpack(&obligation_info.data.borrow())?;
+    if obligation_info.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &obligation.borrow_reserve != borrow_reserve_info.key {
+        msg!("Borrow reserve input doesn't match existing obligation borrow reserve");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &obligation.collateral_reserve != deposit_reserve_info.key {
+        msg!("Collateral reserve input doesn't match existing obligation collateral reserve");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+
+    unpack_mint(&obligation_token_mint_info.data.borrow())?;
+    if &obligation.token_mint != obligation_token_mint_info.key {
+        msg!("Obligation token mint input doesn't match existing obligation token mint");
+        return Err(LendingError::InvalidTokenMint.into());
+    }
+
+    let obligation_token_output = Token::unpack(&obligation_token_output_info.data.borrow())?;
+    if obligation_token_output_info.owner != token_program_id.key {
+        return Err(LendingError::InvalidTokenOwner.into());
+    }
+    if &obligation_token_output.mint != obligation_token_mint_info.key {
+        return Err(LendingError::InvalidTokenMint.into());
+    }
+
     // accrue interest and update rates
     borrow_reserve.accrue_interest(clock.slot)?;
-    deposit_reserve.accrue_interest(clock.slot)?;
```

Comment:
What is the impact of not accruing interest on the deposit reserve? We need the deposit reserve for the trade simulation in order to convert between collateral tokens and liquidity tokens.  Based on our exchange in the calc PR, I thought the exchange rate between collateral tokens and liquidity tokens depended on the interest accrued on the reserve.  Digging a bit deeper, it only depends on the amount of tokens issued, which makes sense, since collateral tokens should NOT grow in value just because time has passed -- there need to be borrows / repays / liquidations.  Anyway, this was a bit of a brain dump, but either way, could you add a comment explaining why it isn't needed?

Reply:
Assuming the worst case of 1 year of interest accrual calculation, not accruing saves about 30k instructions. And cutting out obligation initialization saved about 10k from removing CPI calls.

> I thought the exchange rate between collateral tokens and liquidity tokens depended on the interest accrued on the reserve.

This is indeed the case.

> since collateral tokens should NOT grow in value just because time has passed -- there need to be borrows / repays / liquidations

The way the protocol is implemented allows lenders to continuously earn interest on their lent liquidity. Then, the only trick is having liquidity available to withdraw if they want to cash out. This is where the variable interest rate helps because as the interest rate increases, it incentivizes borrowers to pay back loans and new lenders to add liquidity.

My thinking was that since collateral value is only going up, the worst case with removing the interest accrual is that the loan is underwritten in a way that undervalues collateral, which leads to slightly over-collateralized positions, which didn't seem like an awful thing.

But now I'm thinking it's better to just take the computation hit of accruing the interest to give the client a better experience. In practice, I expect that the reserves will be touched enough that the amortized cost of accruing interest is pretty stable and low.